Agent compatibility issue with Scala 2.13x and Java 11+

## Description
The agent throws exceptions when running on Java 11+ and Scala 2.13x.  
Some customers report interference with reporting Transactions and custom attributes.  

From local reproduction test:

```
2023-04-24T10:35:58,265-0700 [27147 146] com.newrelic FINE: Unable to transform class scala/concurrent/impl/Promise$Transformation
java.lang.TypeNotPresentException: Type scala/util/Try not present
	at com.newrelic.agent.deps.org.objectweb.asm.ClassWriter.getCommonSuperClass(ClassWriter.java:1041) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.SymbolTable.addMergedType(SymbolTable.java:1202) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.Frame.merge(Frame.java:1299) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.Frame.merge(Frame.java:1244) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1611) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1547) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.ClassReader.readCode(ClassReader.java:2665) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1514) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.ClassReader.accept(ClassReader.java:744) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.deps.org.objectweb.asm.ClassReader.accept(ClassReader.java:424) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.instrumentation.custom.ScalaTraitFinalFieldTransformer.doTransform(ScalaTraitFinalFieldTransformer.java:40) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.instrumentation.custom.ScalaTraitFinalFieldTransformer.transform(ScalaTraitFinalFieldTransformer.java:20) ~[newrelic.jar:8.0.1]
	at com.newrelic.agent.instrumentation.context.InstrumentationClassTransformer.transform(InstrumentationClassTransformer.java:144) ~[newrelic.jar:8.0.1]
	at java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:246) ~[?:?]
	at sun.instrument.TransformerManager.transform(TransformerManager.java:188) ~[?:?]
	at sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:563) ~[?:?]
	at java.lang.ClassLoader.defineClass1(Native Method) ~[?:?]
	at java.lang.ClassLoader.defineClass(ClassLoader.java:1017) ~[?:?]
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174) ~[?:?]
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:555) ~[?:?]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:458) ~[?:?]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:452) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
	at java.net.URLClassLoader.findClass(URLClassLoader.java:451) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:589) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?]
	at scala.concurrent.impl.Promise$.<clinit>(Promise.scala:403) ~[scala-library.jar:?]
	at scala.concurrent.impl.Promise$DefaultPromise.<init>(Promise.scala:108) ~[scala-library.jar:?]
	at scala.concurrent.Future$.fromTry(Future.scala:658) ~[scala-library.jar:?]
	at scala.concurrent.Future$.<clinit>(Future.scala:552) ~[scala-library.jar:?]
	at scala.tools.nsc.backend.jvm.GeneratedClassHandler$WritingClassHandler.postProcessUnit(GeneratedClassHandler.scala:126) ~[?:?]
	at scala.tools.nsc.backend.jvm.GeneratedClassHandler$WritingClassHandler.process(GeneratedClassHandler.scala:119) ~[?:?]
	at scala.tools.nsc.backend.jvm.CodeGen.genUnit(CodeGen.scala:68) ~[?:?]
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase.apply(GenBCode.scala:74) ~[?:?]
	at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:467) ~[?:?]
	at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:414) ~[?:?]
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase.super$run(GenBCode.scala:80) ~[?:?]
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase.$anonfun$run$1(GenBCode.scala:80) ~[?:?]
	at scala.tools.nsc.backend.jvm.GenBCode$BCodePhase.run(GenBCode.scala:78) ~[?:?]
	at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1530) ~[?:?]
	at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1514) ~[?:?]
	at scala.tools.nsc.Global$Run.compileSources(Global.scala:1506) ~[?:?]
	at scala.tools.nsc.Global$Run.compile(Global.scala:1640) ~[?:?]
	at scala.tools.nsc.Driver.doCompile(Driver.scala:48) ~[?:?]
	at scala.tools.nsc.MainClass.doCompile(Main.scala:30) ~[?:?]
	at scala.tools.nsc.Driver.process(Driver.scala:68) ~[?:?]
	at scala.tools.nsc.Main.process(Main.scala) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at sbt.internal.inc.RawCompiler.getReporter$1(RawCompiler.scala:56) ~[?:?]
	at sbt.internal.inc.RawCompiler.apply(RawCompiler.scala:77) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$7(AnalyzingCompiler.scala:457) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler$.handleCompilationError$1(AnalyzingCompiler.scala:432) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$5(AnalyzingCompiler.scala:453) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$5$adapted(AnalyzingCompiler.scala:448) ~[?:?]
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:490) ~[?:?]
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:500) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$2(AnalyzingCompiler.scala:448) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$2$adapted(AnalyzingCompiler.scala:440) ~[?:?]
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:490) ~[?:?]
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:500) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler$.compileSources(AnalyzingCompiler.scala:440) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$3(ZincComponentCompiler.scala:275) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$3$adapted(ZincComponentCompiler.scala:257) ~[?:?]
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:490) ~[?:?]
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:500) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$2(ZincComponentCompiler.scala:257) ~[?:?]
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23) ~[scala-library.jar:?]
	at sbt.internal.util.BufferedLogger.bufferQuietly(BufferedLogger.scala:159) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$1(ZincComponentCompiler.scala:257) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$1$adapted(ZincComponentCompiler.scala:254) ~[?:?]
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:490) ~[?:?]
	at sbt.io.IO$.withTemporaryDirectory(IO.scala:500) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler.compileAndInstall(ZincComponentCompiler.scala:254) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compiledBridgeJar$1(ZincComponentCompiler.scala:222) ~[?:?]
	at sbt.internal.inc.IfMissing$Define.run(IfMissing.scala:19) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.createAndCache$1(ZincComponentManager.scala:51) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.$anonfun$files$3(ZincComponentManager.scala:62) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.getOrElse$1(ZincComponentManager.scala:43) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.$anonfun$files$2(ZincComponentManager.scala:62) ~[?:?]
	at sbt.internal.inc.ZincComponentManager$$anon$1.call(ZincComponentManager.scala:91) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:113) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.withChannelRetries$1(Locks.scala:91) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.$anonfun$withFileLock$1(Locks.scala:119) ~[?:?]
	at xsbt.boot.Using$.withResource(Using.scala:12) ~[?:?]
	at xsbt.boot.Using$.apply(Using.scala:9) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.withFileLock(Locks.scala:119) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:71) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:59) ~[?:?]
	at xsbt.boot.Locks$.apply0(Locks.scala:47) ~[?:?]
	at xsbt.boot.Locks$.apply(Locks.scala:36) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.lock(ZincComponentManager.scala:91) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.$anonfun$lockSecondaryCache$1(ZincComponentManager.scala:88) ~[?:?]
	at scala.Option.map(Option.scala:230) ~[scala-library.jar:?]
	at sbt.internal.inc.ZincComponentManager.lockSecondaryCache(ZincComponentManager.scala:88) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.fromSecondary$1(ZincComponentManager.scala:60) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.$anonfun$files$6(ZincComponentManager.scala:66) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.getOrElse$1(ZincComponentManager.scala:43) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.$anonfun$files$5(ZincComponentManager.scala:66) ~[?:?]
	at sbt.internal.inc.ZincComponentManager$$anon$1.call(ZincComponentManager.scala:91) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:113) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.withChannelRetries$1(Locks.scala:91) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.$anonfun$withFileLock$1(Locks.scala:119) ~[?:?]
	at xsbt.boot.Using$.withResource(Using.scala:12) ~[?:?]
	at xsbt.boot.Using$.apply(Using.scala:9) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.withFileLock(Locks.scala:119) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:71) ~[?:?]
	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:59) ~[?:?]
	at xsbt.boot.Locks$.apply0(Locks.scala:47) ~[?:?]
	at xsbt.boot.Locks$.apply(Locks.scala:36) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.lock(ZincComponentManager.scala:91) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.lockLocalCache(ZincComponentManager.scala:84) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.files(ZincComponentManager.scala:66) ~[?:?]
	at sbt.internal.inc.ZincComponentManager.file(ZincComponentManager.scala:72) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler.compiledBridgeJar(ZincComponentCompiler.scala:222) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler$ZincCompilerBridgeProvider.compiledBridge(ZincComponentCompiler.scala:63) ~[?:?]
	at sbt.internal.inc.ZincComponentCompiler$ZincCompilerBridgeProvider.fetchCompiledBridge(ZincComponentCompiler.scala:70) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler.getDualLoader(AnalyzingCompiler.scala:354) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler.getCompilerLoader(AnalyzingCompiler.scala:343) ~[?:?]
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:87) ~[?:?]
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193) ~[?:?]
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23) ~[scala-library.jar:?]
	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248) ~[?:?]
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183) ~[?:?]
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163) ~[?:?]
	at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239) ~[?:?]
	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163) ~[?:?]
	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211) ~[?:?]
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534) ~[?:?]
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534) ~[?:?]
	at sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:179) ~[?:?]
	at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:177) ~[?:?]
	at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:463) ~[?:?]
	at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116) ~[?:?]
	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56) ~[?:?]
	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52) ~[?:?]
	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263) ~[?:?]
	at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:418) ~[?:?]
	at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506) ~[?:?]
	at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:405) ~[?:?]
	at sbt.internal.inc.Incremental$.apply(Incremental.scala:171) ~[?:?]
	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534) ~[?:?]
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488) ~[?:?]
	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332) ~[?:?]
	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425) ~[?:?]
	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137) ~[?:?]
	at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2363) ~[?:?]
	at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2313) ~[?:?]
	at sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:30) ~[?:?]
	at sbt.internal.io.Retry$.apply(Retry.scala:46) ~[?:?]
	at sbt.internal.io.Retry$.apply(Retry.scala:28) ~[?:?]
	at sbt.internal.io.Retry$.apply(Retry.scala:23) ~[?:?]
	at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:30) ~[?:?]
	at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2311) ~[?:?]
	at scala.Function1.$anonfun$compose$1(Function1.scala:49) ~[scala-library.jar:?]
	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62) ~[collections_2.12-1.7.2.jar:1.7.2]
	at sbt.std.Transform$$anon$4.work(Transform.scala:68) ~[task-system_2.12-1.7.2.jar:1.7.2]
	at sbt.Execute.$anonfun$submit$2(Execute.scala:282) ~[tasks_2.12-1.7.2.jar:1.7.2]
	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23) ~[util-control_2.12-1.7.2.jar:1.7.2]
	at sbt.Execute.work(Execute.scala:291) ~[tasks_2.12-1.7.2.jar:1.7.2]
	at sbt.Execute.$anonfun$submit$1(Execute.scala:282) ~[tasks_2.12-1.7.2.jar:1.7.2]
	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265) ~[tasks_2.12-1.7.2.jar:1.7.2]
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64) [tasks_2.12-1.7.2.jar:1.7.2]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.ClassNotFoundException: scala.util.Try
	at java.net.URLClassLoader.findClass(URLClassLoader.java:476) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:589) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?]
	at java.lang.Class.forName0(Native Method) ~[?:?]
	at java.lang.Class.forName(Class.java:398) ~[?:?]
	at com.newrelic.agent.deps.org.objectweb.asm.ClassWriter.getCommonSuperClass(ClassWriter.java:1039) ~[newrelic.jar:8.0.1]
	... 170 more

```

## Expected Behavior
No exceptions thrown and agent otherwise reports all expected transactions and attributes.
## Steps to Reproduce
This can be reproduced:
Quick and easy: pull this app down and setup to run with the agent as described in the readme.
https://github.com/newrelic/newrelic-java-examples/tree/main/newrelic-java-agent/scala/segment-api-synchronous
Update sbt build and .jvmoptions to the versions given below.
Run the app.

This app uses the scala api (which essentially just calls the java agent api).  However, you can change the example app to use the direct Java API calls like so:

```

  @Trace
  def trace1: Unit ={
    val i = 1
    val j = 2
    println(i + j)
  }

  @Trace
  def trace2: Unit ={
    NewRelic.addCustomParameter("key", "value")
    val i = 1
    val j = 2
    println(i + j)
  }

  @Trace(dispatcher=true)
  def startTransaction: Unit ={
    trace1
    trace2
  }

  startTransaction

```
(then update the libraryDependencies to accomodate)

## Your Environment
Java 11 zulu hotspot.  Scala 2.13.10.

--- a/newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/custom/ScalaTraitFinalFieldTransformer.java
+++ b/newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/custom/ScalaTraitFinalFieldTransformer.java
@@ -3,6 +3,7 @@ package com.newrelic.agent.instrumentation.custom;
 import com.newrelic.agent.instrumentation.classmatchers.OptimizedClassMatcher;
 import com.newrelic.agent.instrumentation.context.ContextClassTransformer;
 import com.newrelic.agent.instrumentation.context.InstrumentationContext;
+import com.newrelic.agent.util.asm.PatchedClassWriter;
 import org.objectweb.asm.ClassReader;
 import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.ClassWriter;
@@ -34,7 +35,7 @@ public class ScalaTraitFinalFieldTransformer implements ContextClassTransformer 
 
     LOG.debug("Instrumenting class " + className);
     ClassReader reader = new ClassReader(classfileBuffer);
-    ClassWriter writer = new ClassWriter(ClassWriter.COMPUTE_FRAMES);
+    ClassWriter writer = new PatchedClassWriter(ClassWriter.COMPUTE_FRAMES, loader);
     ClassVisitor cv = writer;
     cv = new ScalaTraitFinalFieldVisitor(cv, context.getScalaFinalFields());
     reader.accept(cv, ClassReader.SKIP_FRAMES);

--- a/newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/custom/ScalaTraitFinalFieldTransformer.java
+++ b/newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/custom/ScalaTraitFinalFieldTransformer.java
@@ -3,6 +3,7 @@ package com.newrelic.agent.instrumentation.custom;
 import com.newrelic.agent.instrumentation.classmatchers.OptimizedClassMatcher;
 import com.newrelic.agent.instrumentation.context.ContextClassTransformer;
 import com.newrelic.agent.instrumentation.context.InstrumentationContext;
+import com.newrelic.agent.util.asm.CustomClassLoaderClassWriter;
 import com.newrelic.agent.util.asm.PatchedClassWriter;
 import org.objectweb.asm.ClassReader;
 import org.objectweb.asm.ClassVisitor;
@@ -35,7 +36,7 @@ public class ScalaTraitFinalFieldTransformer implements ContextClassTransformer 
 
     LOG.debug("Instrumenting class " + className);
     ClassReader reader = new ClassReader(classfileBuffer);
-    ClassWriter writer = new PatchedClassWriter(ClassWriter.COMPUTE_FRAMES, loader);
+    ClassWriter writer = new CustomClassLoaderClassWriter(ClassWriter.COMPUTE_FRAMES, loader);
     ClassVisitor cv = writer;
     cv = new ScalaTraitFinalFieldVisitor(cv, context.getScalaFinalFields());
     reader.accept(cv, ClassReader.SKIP_FRAMES);

--- a/newrelic-agent/src/main/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriter.java
+++ b/newrelic-agent/src/main/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriter.java
@@ -2,6 +2,7 @@ package com.newrelic.agent.util.asm;
 
 import com.newrelic.agent.Agent;
 import com.newrelic.agent.bridge.AgentBridge;
+import org.objectweb.asm.ClassReader;
 
 import java.io.IOException;
 import java.util.logging.Level;
@@ -32,7 +33,10 @@ public class CustomClassLoaderClassWriter extends PatchedClassWriter {
                 Agent.LOG.log(Level.FINEST, "class not found in base classloader: "+type);
                 try {
                     // if all else fails, let's try the hard way
-                    result = getClassReader(type).getClass();
+                    // this case exists because of continued TypeNotPresentExceptions when instrumenting Scala
+                    ClassReader classReader = getClassReader(type);
+                    if (classReader == null) throw new ClassNotFoundException("ClassReader not found for class: "+type);
+                    result = classReader.getClass();
                 } catch (IOException ioe) {
                     Agent.LOG.log(Level.FINEST, ioe.toString(), ioe);
                     throw new ClassNotFoundException("Could not find class via ClassReader: "+type);

--- a/newrelic-agent/src/main/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriter.java
+++ b/newrelic-agent/src/main/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriter.java
@@ -1,0 +1,76 @@
+package com.newrelic.agent.util.asm;
+
+import com.newrelic.agent.Agent;
+import com.newrelic.agent.bridge.AgentBridge;
+
+import java.io.IOException;
+import java.util.logging.Level;
+
+public class CustomClassLoaderClassWriter extends PatchedClassWriter {
+
+    private final ClassLoader classLoader;
+    private final ClassResolver classResolver;
+
+    public CustomClassLoaderClassWriter(int flags, ClassLoader classLoader) {
+        super(flags, classLoader);
+        this.classLoader = classLoader;
+        this.classResolver = ClassResolvers.getClassLoaderResolver(classLoader == null ?
+                AgentBridge.getAgent().getClass().getClassLoader() : classLoader);
+    }
+
+    private Class loadClass(String type) throws ClassNotFoundException {
+        Class result = null;
+        try {
+            // try the custom classloader first
+            result = classLoader.loadClass(type);
+        } catch (ClassNotFoundException e) {
+            Agent.LOG.log(Level.FINEST, "class not found in custom classloader: "+type);
+            try {
+                // now try the base classloader
+                result = this.getClass().getClassLoader().loadClass(type);
+            } catch (ClassNotFoundException e2) {
+                Agent.LOG.log(Level.FINEST, "class not found in base classloader: "+type);
+                try {
+                    // if all else fails, let's try the hard way
+                    result = getClassReader(type).getClass();
+                } catch (IOException ioe) {
+                    Agent.LOG.log(Level.FINEST, ioe.toString(), ioe);
+                    throw new ClassNotFoundException("Could not find class via ClassReader: "+type);
+                }
+            }
+        }
+
+        return result;
+    }
+
+    protected String getCommonSuperClass(String type1, String type2) {
+        Class class1;
+        try {
+            class1 = loadClass(type1);
+        } catch (ClassNotFoundException e) {
+            throw new TypeNotPresentException(type1, e);
+        }
+
+        Class class2;
+        try {
+            class2 = loadClass(type2);
+        } catch (ClassNotFoundException e) {
+            throw new TypeNotPresentException(type2, e);
+        }
+
+        if (class1.isAssignableFrom(class2)) {
+            return type1;
+        } else if (class2.isAssignableFrom(class1)) {
+            return type2;
+        } else if (!class1.isInterface() && !class2.isInterface()) {
+            do {
+                class1 = class1.getSuperclass();
+            } while(!class1.isAssignableFrom(class2));
+
+            return class1.getName().replace('.', '/');
+        } else {
+            return "java/lang/Object";
+        }
+    }
+
+}

--- a/newrelic-agent/src/main/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriter.java
+++ b/newrelic-agent/src/main/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriter.java
@@ -35,8 +35,9 @@ public class CustomClassLoaderClassWriter extends PatchedClassWriter {
                     // if all else fails, let's try the hard way
                     // this case exists because of continued TypeNotPresentExceptions when instrumenting Scala
                     ClassReader classReader = getClassReader(type);
-                    if (classReader == null) throw new ClassNotFoundException("ClassReader not found for class: "+type);
-                    result = classReader.getClass();
+                    if (classReader != null) {
+                        result = classReader.getClass();
+                    }
                 } catch (IOException ioe) {
                     Agent.LOG.log(Level.FINEST, ioe.toString(), ioe);
                     throw new ClassNotFoundException("Could not find class via ClassReader: "+type);
@@ -60,6 +61,10 @@ public class CustomClassLoaderClassWriter extends PatchedClassWriter {
             class2 = loadClass(type2);
         } catch (ClassNotFoundException e) {
             throw new TypeNotPresentException(type2, e);
+        }
+
+        if (class1 == null || class2 == null) {
+            return JAVA_LANG_OBJECT;
         }
 
         if (class1.isAssignableFrom(class2)) {

--- a/newrelic-agent/src/main/java/com/newrelic/agent/util/asm/PatchedClassWriter.java
+++ b/newrelic-agent/src/main/java/com/newrelic/agent/util/asm/PatchedClassWriter.java
@@ -29,7 +29,7 @@ import java.util.logging.Level;
  * structure without actually loading any classes.
  */
 public class PatchedClassWriter extends ClassWriter {
-    private static final String JAVA_LANG_OBJECT = "java/lang/Object";
+    static final String JAVA_LANG_OBJECT = "java/lang/Object";
     protected final ClassResolver classResolver;
 
     public PatchedClassWriter(int flags, ClassLoader classLoader) {

--- a/newrelic-agent/src/test/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriterTest.java
+++ b/newrelic-agent/src/test/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriterTest.java
@@ -1,0 +1,53 @@
+package com.newrelic.agent.util.asm;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.objectweb.asm.ClassWriter;
+
+import java.io.Serializable;
+
+public class CustomClassLoaderClassWriterTest {
+
+    CustomClassLoaderClassWriter api = new CustomClassLoaderClassWriter(ClassWriter.COMPUTE_FRAMES, this.getClass().getClassLoader());
+
+    @Test
+    public void test_getCommonSuperClass_sameClass() {
+        String className = CustomClassLoaderClassWriterTest.class.getName();
+        Assert.assertEquals(className, api.getCommonSuperClass(className, className));
+    }
+
+    @Test
+    public void test_getCommonSuperClass_commonSuper() {
+        Assert.assertEquals(Object.class.getName().replace('.','/'),
+                api.getCommonSuperClass(String.class.getName(), Integer.class.getName()));
+    }
+
+    @Test
+    public void test_getCommonSuperClass_commonSuper2() {
+        Assert.assertEquals(Number.class.getName(),
+                api.getCommonSuperClass(Integer.class.getName(), Number.class.getName()));
+    }
+
+    @Test
+    public void test_getCommonSuperClass_interfaces() {
+        Assert.assertEquals(Object.class.getName().replace('.','/'),
+                api.getCommonSuperClass(Comparable.class.getName(), Serializable.class.getName()));
+    }
+
+    @Test
+    public void test_getCommonSuperClass_noCommonSuper() {
+        Assert.assertEquals(Object.class.getName().replace('.','/'),
+                api.getCommonSuperClass(String.class.getName(), CustomClassLoaderClassWriterTest.class.getName()));
+    }
+
+    @Test (expected = TypeNotPresentException.class)
+    public void test_getCommonSuperClass_invalidClass1() {
+        api.getCommonSuperClass("noexist1", CustomClassLoaderClassWriterTest.class.getName());
+    }
+
+    @Test (expected = TypeNotPresentException.class)
+    public void test_getCommonSuperClass_invalidClass2() {
+        api.getCommonSuperClass(CustomClassLoaderClassWriterTest.class.getName(), "noexist2");
+    }
+
+}

--- a/newrelic-agent/src/test/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriterTest.java
+++ b/newrelic-agent/src/test/java/com/newrelic/agent/util/asm/CustomClassLoaderClassWriterTest.java
@@ -40,14 +40,16 @@ public class CustomClassLoaderClassWriterTest {
                 api.getCommonSuperClass(String.class.getName(), CustomClassLoaderClassWriterTest.class.getName()));
     }
 
-    @Test (expected = TypeNotPresentException.class)
+    @Test
     public void test_getCommonSuperClass_invalidClass1() {
-        api.getCommonSuperClass("noexist1", CustomClassLoaderClassWriterTest.class.getName());
+        Assert.assertEquals(Object.class.getName().replace('.', '/'),
+                api.getCommonSuperClass("noexist1", CustomClassLoaderClassWriterTest.class.getName()));
     }
 
-    @Test (expected = TypeNotPresentException.class)
+    @Test
     public void test_getCommonSuperClass_invalidClass2() {
-        api.getCommonSuperClass(CustomClassLoaderClassWriterTest.class.getName(), "noexist2");
+        Assert.assertEquals(Object.class.getName().replace('.', '/'),
+                api.getCommonSuperClass(CustomClassLoaderClassWriterTest.class.getName(), "noexist2"));
     }
 
 }

